hadoop-common-project/hadoop-common/src/main/java/org/apache/hadoop/ipc/CallerContext.java

```diff
@@ -114,114 +120,160 @@ public String toString() {
-     * The illegal separators include '\t', '\n', '='.
-     * Throw IllegalArgumentException if the separator is Illegal.
-     * @param separator the separator of fields.
-     */
-    private void checkFieldSeparator(String separator) {
-      if (ILLEGAL_SEPARATORS.contains(separator)) {
-        throw new IllegalArgumentException("Illegal field separator: "
-            + separator);
+  /** The caller context builder. */
+  public static final class Builder {
+
+    private byte[] signature;
+    private final Configs configs;
+    private final StringBuilder ctx = new StringBuilder();
+
+    public Builder(String context) {
+      this(context, null);
+    }
+
+    public Builder(String context, Configuration conf) {
+      if (isValidFiled(context)) {
+        ctx.append(context);
+      }
+      configs = Configs.getInstance(conf);
+    }
+
+    @VisibleForTesting
+    Builder(String context, Configuration conf, boolean refresh) {
+      this(context, null);
+      if (refresh) {
+        configs.init(conf); // refresh configs for testing
       }
     }
 
     /**
      * Whether the field is valid.
-     * @param field one of the fields in context.
+     * @param field target field.
      * @return true if the field is not null or empty.
      */
-    private boolean isValid(String field) {
+    private boolean isValidFiled(String field) {
       return field != null && field.length() > 0;
     }
 
     public Builder setSignature(byte[] signature) {
-      if (signature != null && signature.length > 0) {
+      if (isSignatureValid(signature)) {
         this.signature = Arrays.copyOf(signature, signature.length);
       }
       return this;
     }
 
     /**
-     * Get the context.
-     * For example, the context is "key1:value1,key2:value2".
-     * @return the valid context or null.
-     */
-    public String getContext() {
-      return sb.length() > 0 ? sb.toString() : null;
-    }
-
-    /**
-     * Get the signature.
-     * @return the signature.
+     * Whether the signature is valid.
+     * @param signature target signature.
+     * @return true if the signature is not null and the length of signature
+     *         is greater than 0 and less than or equal to the length limit.
      */
-    public byte[] getSignature() {
-      return signature;
+    private boolean isSignatureValid(byte[] signature) {
+      return signature != null && signature.length > 0
+          && signature.length <= configs.signatureMaxLen;
     }
 
     /**
-     * Append new field to the context.
+     * Append new field to the context if the filed is not empty and length
+     * of the context is not exceeded limit.
      * @param field one of fields to append.
      * @return the builder.
      */
     public Builder append(String field) {
-      if (isValid(field)) {
-        if (sb.length() > 0) {
-          sb.append(fieldSeparator);
+      if (isValidFiled(field) && !isContextLengthExceedLimit()) {
```

Comment:
How about logging a warning message here or throw an exception? Makes troubleshooting easier.

Reply:
> How about logging a warning message here or throw an exception? Makes troubleshooting easier.
 Thanks for review, I think logging a warning message seems better, I will add it soon.